modules/renter/contractor/contractor.go

```diff
@@ -41,27 +41,29 @@ type Contractor struct {
 	hdb     hostDB
 	log     *persist.Logger
 	persist persister
+	mu      sync.RWMutex
 	tpool   transactionPool
 	wallet  wallet
 
-	allowance       modules.Allowance
-	blockHeight     types.BlockHeight
+	// in addition to mu, a separate lock enforces that multiple goroutines
+	// won't try to simultaneously edit the contract set.
+	editLock siasync.TryMutex
+
+	allowance     modules.Allowance
+	blockHeight   types.BlockHeight
+	currentPeriod types.BlockHeight
+	lastChange    modules.ConsensusChangeID
+
+	downloaders map[types.FileContractID]*hostDownloader
+	editors     map[types.FileContractID]*hostEditor
+	renewing    map[types.FileContractID]bool // prevent revising during renewal
+	revising    map[types.FileContractID]bool // prevent overlapping revisions
+
 	cachedRevisions map[types.FileContractID]cachedRevision
 	contracts       map[types.FileContractID]modules.RenterContract
-	currentPeriod   types.BlockHeight
-	downloaders     map[types.FileContractID]*hostDownloader
-	editors         map[types.FileContractID]*hostEditor
-	lastChange      modules.ConsensusChangeID
 	oldContracts    map[types.FileContractID]modules.RenterContract
+	relationships   map[types.FileContractID]types.SiaPublicKey
```

Comment:
I determined that the compatibility issues would be harder to work around.

This map is temporary, just like the oldContracts map. I'm procedurally migrating everything to just using pubkeys, as we do this in more parts of the codebase we may accumulate a few more helper-maps, but then finally we'll be able to drop them all at the same time.